src/java.base/share/classes/java/nio/Heap-X-Buffer.java.template

```diff
@@ -278,6 +278,63 @@ class Heap$Type$Buffer$RW$
 
 #if[char]
 
+    //
+    // Use getChars() to load chars directly into the heap buffer array.
+    // For a String or StringBuffer source this improves performance if
+    // a proper subsequence is being appended as copying to a new intermediate
+    // String object is avoided. For a StringBuilder where either a subsequence
+    // or the full sequence of chars is being appended, copying the chars to
+    // an intermedite String in StringBuilder::toString is avoided.
+    //
+    private $Type$Buffer appendChars(CharSequence csq, int start, int end) {
+        checkSession();
+
+        int length = end - start;
+        int pos = position();
+        int lim = limit();
+        int rem = (pos <= lim) ? lim - pos : 0;
+        if (length > rem)
+            throw new BufferOverflowException();
+
+        if (csq instanceof StringBuilder bld) {
+            bld.getChars(start, end, hb, ix(pos));
+        } else if (csq instanceof String str) {
+            str.getChars(start, end, hb, ix(pos));
+        } else if (csq instanceof StringBuffer buf) {
+            buf.getChars(start, end, hb, ix(pos));
+        }
```

Comment:
Surprised to see legacy StringBuffer here but okay. A small suggestion is to keep check for String first, and keep the naming from the previous iteration, i.e.

```
        if (csq instanceof String str) {
            str.getChars(start, end, hb, ix(pos));
        } else if (csq instanceof StringBuilder sb) {
            sb.getChars(start, end, hb, ix(pos));
        } else if (csq instanceof StringBuffer sb) {
            sb.getChars(start, end, hb, ix(pos));
        }
```

Reply:
I put `StringBuilder` first as it is always used for the `append` without `start` and `end` and sometimes for the version without but I've no problem changing it.